Homework 22 POM

anatestpro opened the pull request:
Not sure why it gives me an error. Please help me to fix it. Thank you!!

anatestpro committed:
Delete Playlist Test

Please help me to fix an error. For some reason I have problems with variables url and driver. I tried to identify them, but couldn't for some reason. What's the problem? Thank you for providing the solution!

anatestpro committed:
Homework19 - Delete Playlist Test

anatestpro committed:
Homework20

Delete Playlist Test with Explicit waits

anatestpro committed:
HW 17 Adding a song to the playlist test

anatestpro committed:
Homework21 Rename Playlist test

anatestpro committed:
homework22 POM

Please help me to correct an error. Not sure what is wrong. Thank you!!

rajniss21 commented:
Hello @anatestpro,
I have checked your homework 22, but you got this error in your codebase, Please fix your code.
Best wishes from the HW team (Rajnish, rajnish@testpro.io)!


1. Please refactor your class name from Homework22.java to Homework22.
2. Add your test case in a similar way as you have defined in Homework21.
3. Class names are generally written with Capital letters, please follow the naming convention.

![Screenshot_79](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/988ffd33-58d7-4f96-b9f0-e93307d934f4)

anatestpro commented:
Hi! I was speaking about errors in pages section. Not sure how to fix it. Could you please take a look?

![Screenshot 2023-08-14 084236](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/139082811/3051c6cf-4b63-4700-9006-4d1df177b488)

anatestpro committed:
hw22

rajniss21 commented:
Hello @anatestpro,

The issue with your codebase is the following:

1. It appears that there are two instances of the LoginTests classes, leading to a conflict. Please remove the unused LoginTests class to resolve this issue.
2. The LoginTests class is attempting to extend the BaseTest from a location outside the directory's scope. To address this, move the definition of the LoginTests class outside the Page directory.

Hope this may help. if you face any issues. Please leave a comment and I'll reach out to you.

![Screenshot_80](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/50fc57a6-f0b5-4bd0-afc9-9682a5602d5b)
![Screenshot_82](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/c10a52e2-6e55-4e74-9bda-e11d15e4a1d9)
![Screenshot_85](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/88a08db7-6220-45fa-8350-8a6452e3a971)
![Screenshot_84](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/54f622f7-aa60-4ea6-afdf-1098ce94ff12)
![Screenshot_83](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/ee9a73c6-be0e-413d-bbb6-39a5239174a8)
![Screenshot_81](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/25093177/45cef78b-d555-45ad-ace1-437077c46bc7)

anatestpro commented:
Hi! I really need your help. My tests stopped running since I updated my chrome browser. I've tried so many different methods (like changing webdriver version and selenium and testNG version in build.gradle). When I changed versions, it leads to maven files. I see so many recommendations regarding pom.xml and JUnit, but can't find much about gradle and testNG. I've spent three days trying to figure out. And I'm hopeless at this point. Please help me to figure it out.

It gives me this error: org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: session not created: This version of ChromeDriver only supports Chrome version 114
Current browser version is 122.0.6261.94 with binary path /Applications/Google Chrome.app/Contents/MacOS/Google Chrome 
Host info: host: 'Anastasias-MBP.socal.rr.com', ip: '2603:8000:6640:c6:6800:3e6:8d4d:dfd9%en0'
Build info: version: '4.18.1', revision: 'b1d3319b48'
System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '14.2.1', java.version: '17.0.9'
Driver info: org.openqa.selenium.chrome.ChromeDriver
<img width="1427" alt="Screenshot 2024-03-03 at 11 28 30 PM" src="https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/139082811/302c7f3c-a0ef-4c47-b6d9-4c7ea194d557">
<img width="1389" alt="Screenshot 2024-03-03 at 11 25 43 PM" src="https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/139082811/bbf2e55b-00d8-455f-9e80-ba443f63bdab">

Berichicko commented on src/test/java/pages/BasePage.java:
```diff
@@ -0,0 +1,28 @@
+import java.time.Duration;
+
+public class BasePage {
+    public static WebDriver driver= null;
+    public static WebDriverWait wait = null;
+    public static Actions actions = null;
+    public static String url = null;
```
I recommend: make the variables protected, because protected access gives the subclass a chance to use the helper method or variable, while preventing a nonrelated class from trying to use it.

```
    protected Actions actions;
    protected WebDriverWait wait;
    protected WebDriver driver;
```

Berichicko commented on src/test/java/pages/LoginPage.java:
```diff
@@ -0,0 +1,33 @@
+        super(givenDriver);
+    }
+
+    By emailField = By.cssSelector("[type='email']");
+    By passwordField = By.cssSelector("[type='password']");
+    By submitBtn = By.cssSelector("[type='submit']");
```
You should use encapsulation of non-static and non-final fields.

If a variable is used only by other methods of that class, it should be declared as private.

I recommend:  make all the locators and WebElements private, because private keyword in java allows most restrictive access to variables and methods and offer strongest form of Encapsulation. 

For example

    ```
    private By shuffleBtnLocator = By.cssSelector(".btn-shuffle-all");
    private By firstSongLocator = By.cssSelector(".play");

```

Berichicko commented:
The code is not compiled! Please finish the homework!You got this problem, please correct it.

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/feccd2bf-3f4b-48fa-b876-e8ec160087a8)

HINT

**Step 1**

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/23a92de6-82ac-4308-ad77-188a1e86d275)

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/e4008231-d9e5-4ffc-a8cf-424fa6c35387)

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/8947daf3-299e-438d-af4f-155e2184c53f)

**Step 2**

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/86e42118-06f4-44b6-b874-3e9f9c8f32bf)

```
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Parameters;

public class BaseTest {

  protected WebDriver driver = null;

  @BeforeSuite
  static void setupClass() {
    WebDriverManager.chromedriver().setup();
  }

  @BeforeMethod
  @Parameters({"BaseURL"})
  public void launchBrowser(String BaseURL)  {
    ChromeOptions options = new ChromeOptions();
    options.addArguments("--disable-notifications", "--remote-allow-origins=*", "--incognito", "--start-maximized", "-lang=en");
    options.setExperimentalOption("excludeSwitches", new String[]{"enable-automation"});
    driver = new ChromeDriver(options);
    driver.get(BaseURL);

  }

  @AfterMethod
  public void closeBrowser() {
    driver.quit();
  }
}
```

**Step 3**

```
import org.testng.annotations.Test;
import pages.LoginPage;

public class HomeWork22 extends BaseTest{

  @Test
  public void registrationNavigation() {
    LoginPage loginPage = new LoginPage(driver);
    loginPage.login();

  }

}
```

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/dc91edbb-7111-41ac-932f-e5f3de63e267)

**Step 4**

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/b353cd6e-8ad1-4851-bb76-a740d6788de9)


```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="TestNG">
    <parameter name="BaseURL" value="https://qa.koel.app/"/>
    <test name="Example Test" preserve-order="false">
        <classes>
            <class name="HomeWork22"/>
        </classes>
    </test>
</suite>
```

**Step 5**

Uptade     `testImplementation 'io.github.bonigarcia:webdrivermanager:5.6.3'`

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/cc2f2c64-9b25-488a-bd4b-9130dbfaddac)

**Step 6**

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/cb7ada45-9b53-4a80-909b-ac348d8962c0)


![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/eb6ab53c-8cb2-40cb-af7e-dd3ebf29004b)

anatestpro commented:
Hi! After changing "testImplementation 'io.github.bonigarcia:webdrivermanager:5.6.3'", I start to have more errors. How do I fix it? THANK YOU!

Same error about not being able to start the new session as I'm using the latest version ChromeDriver 122.0.6261.94but in the error message it says that  114.0.5735.90 is the compatible version. I had before updating my chrome, after update no tests run :( 

<img width="1392" alt="Screenshot 2024-03-04 at 10 34 31 PM" src="https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/139082811/18bb023f-1abb-4c9a-954f-7214afabad30">
<img width="1411" alt="Screenshot 2024-03-04 at 10 34 42 PM" src="https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/139082811/00d7d448-a560-4f04-a85e-0a114ec32c1b">

Berichicko commented:
> Hi! After changing "testImplementation 'io.github.bonigarcia:webdrivermanager:5.6.3'", I start to have more errors. How do I fix it? THANK YOU!
> 
> Same error about not being able to start the new session as I'm using the latest version ChromeDriver 122.0.6261.94but in the error message it says that 114.0.5735.90 is the compatible version. I had before updating my chrome, after update no tests run :(

Hello

Please, re-submit homework for review
